File: 프로그래머스/2/92342. 양궁대회/양궁대회.cpp
#include <bits/stdc++.h>
using namespace std;

int N;
vector<int> apeach(11);
vector<int> best(11, -1);
int maxDiff = -1;

void dfs(int idx, int remain, vector<int>& ryan) {
    if (idx == 11 || remain == 0) {
        // 남은 화살은 0점 칸에 몰아주기
        if (remain > 0) ryan[10] += remain;

        int aScore = 0, rScore = 0;
        for (int i = 0; i < 11; i++) {
            if (apeach[i] == 0 && ryan[i] == 0) continue;
            if (ryan[i] > apeach[i]) rScore += (10 - i);
            else aScore += (10 - i);
        }

        int diff = rScore - aScore;
        if (diff > 0) {
            if (diff > maxDiff) {
                maxDiff = diff;
                best = ryan;
            } else if (diff == maxDiff) {
                // 낮은 점수부터 비교
                for (int i = 10; i >= 0; i--) {
                    if (ryan[i] > best[i]) {
                        best = ryan;
                        break;
                    } else if (ryan[i] < best[i]) break;
                }
            }
        }

        if (remain > 0) ryan[10] -= remain; // 복구
        return;
    }

    // 1. 해당 점수를 이길 수 있을 경우 -> info[idx]+1 발 쏘기
    if (remain > apeach[idx]) {
        ryan[idx] = apeach[idx] + 1;
        dfs(idx + 1, remain - ryan[idx], ryan);
        ryan[idx] = 0; // backtracking
    }

    // 2. 그냥 포기하고 0발 쏘기
    dfs(idx + 1, remain, ryan);
}

vector<int> solution(int n, vector<int> info) {
    N = n;
    apeach = info;
    vector<int> ryan(11, 0);
    dfs(0, N, ryan);

    if (maxDiff <= 0) return {-1};
    return best;
}

int main() {
    vector<int> info = {2,1,1,1,0,0,0,0,0,0,0};
    int n = 5;
    auto ans = solution(n, info);
    for (int x : ans) cout << x << " ";
    cout << endl;
}

File: 프로그래머스/2/92342. 양궁대회/양궁대회.py
from typing import List

max_diff = -1
best: List[int] = []

def dfs(idx: int, arrows: int, apeach: List[int], ryan: List[int]) -> None:
    global max_diff, best
    if idx == 11 or arrows == 0:
        if arrows > 0:
            ryan[10] += arrows
        apeach_score = ryan_score = 0
        for score in range(11):
            if apeach[score] == 0 and ryan[score] == 0:
                continue
            if ryan[score] > apeach[score]:
                ryan_score += 10 - score
            else:
                apeach_score += 10 - score
        diff = ryan_score - apeach_score
        if diff > 0:
            if diff > max_diff:
                max_diff = diff
                best = ryan[:]
            elif diff == max_diff:
                for idx_reverse in range(10, -1, -1):
                    if ryan[idx_reverse] > best[idx_reverse]:
                        best = ryan[:]
                        break
                    if ryan[idx_reverse] < best[idx_reverse]:
                        break
        if arrows > 0:
            ryan[10] -= arrows
        return

    needed = apeach[idx] + 1
    if needed <= arrows:
        ryan[idx] = needed
        dfs(idx + 1, arrows - needed, apeach, ryan)
        ryan[idx] = 0
    dfs(idx + 1, arrows, apeach, ryan)


def solution(n: int, info: List[int]) -> List[int]:
    global max_diff, best
    max_diff = -1
    best = [-1]
    dfs(0, n, info, [0] * 11)
    return best if max_diff > 0 else [-1]
